### PR TITLE
Reorganize Copy task unit tests

### LINKED ISSUE #8767: [Bug]: Build Issue - Unit tests for the Copy task are redundantly executed
### Issue Description

The unit tests for the Copy tasks are in several different class types. The `CopyHardLink_Tests`, `CopyNotHardLink_Tests`, `CopySingleThreaded_Tests`, and `CopySymbolicLink_Tests` classes derive from `Copy_Tests`. All of the tests in the base class also run as part of each derived class. 

### Steps to Reproduce

1. Run unit tests, e.g. `./build.sh --test` or `build.cmd -test`
2. In artifacts/TestResults open the Microsoft.Build.Tasks.UnitTests_*_*.html file(s).
3. Search for the `InvalidRetryCount` test and note that the test was run 5 times.

### Expected Behavior

See Analysis

### Actual Behavior

See Analysis

### Analysis

Class inheritance was used to handle running a given test with different behavior flags. But there is a mixture of tests with different dependencies on the behavior flags. The `InvalidRetryCount` test, as an example, doesn't change with the types of links and threading. Five duplicate executions of an identical test are a waste.

`Xunit.Theory` is a better mechanism for running a given test with different behavior flags.

This issue was found while working on #5881

### Versions & Configurations

_No response_

### PR DESCRIPTION
**Pull request by jrdodds**
Fixes #8767

### Context
I was investigating the Copy task for another issue and I realized that a unit test was being run five times: once for each of the class types `Copy_Tests`, `CopyHardLink_Tests`, `CopyNotHardLink_Tests`, `CopySingleThreaded_Tests`, and `CopySymbolicLink_Tests`.

I found that there is a large set of unit tests for the Copy task that are redundantly executed when tests are run.

Instead of one test class, there is a set of classes for unit tests for the Copy task. The `CopyHardLink_Tests`, `CopyNotHardLink_Tests`, `CopySingleThreaded_Tests`, and `CopySymbolicLink_Tests` classes derive from `Copy_Tests`. All of the tests in the base class also run as part of each derived class. For some tests there are variances in behavior that are controlled by properties acting as flags. The classes that derive from `Copy_Tests`, set the appropriate properties on the base class.

Not all the tests need or use the flags but are effectively re-run regardless.

The following table lists the class types.

| Class | # of Tests | Derives from | Variance from base |
| :---- | ----- | ----- | ----- |
| Copy_Tests | 41 | | |
| CopyHardLink_Tests | 3 (+ base) | Copy_Tests | Hard Link |
| CopyNotHardLink_Tests | 0 (+ base) | Copy_Tests | _none_ |
| CopySingleThreaded_Tests | 0 (+ base) | Copy_Tests | Single Threaded |
| CopySymbolicLink_Tests | 1 (+ base) | Copy_Tests | Sym Link |
| CopyHardAndSymbolicLink_Tests | 3 | | |

The `CopyNotHardLink_Tests` class sets no flags, adds no tests of its own, and effectively just creates duplicate runs of every test in the `Copy_Tests` class.

Adding a new test requires making a choice about where to put the test: in one of the six existing classes or in a new class?

### Changes Made
All tests are moved to one class: `Copy_Tests`. Usage of `Xunit.Theory` replaces class derivation as the means of running the same test for different flags.

### Testing
Tested on Windows 11 and macOS 12. Tested in the VS IDE and by running all unit tests at the command line.

### Notes
This PR shifts from class derivation to `Xunit.Theory` and removes clearly redundant tests. By intent, this PR does not add new tests and does not add new coverage by existing tests.

There are missing coverage cases and there may be further inefficiencies that could be addressed.

### CODE REVIEW

**Comment by JanKrivanek** [comment]
Nice unification and refactoring! Thank you!

One quick question based on high level view - to make sure something is not hidden somewhere:

This is what I see in data for sample CI run prior the change:

0.928s✔ Microsoft.Build.UnitTests.Copy_Tests   (41 tests)
0.036s✔ Microsoft.Build.UnitTests.CopyHardAndSymbolicLink_Tests   (5 tests)
2.667s✔ Microsoft.Build.UnitTests.CopyHardLink_Tests   (45 tests)
1.374s✔ Microsoft.Build.UnitTests.CopyNotHardLink_Tests   (41 tests)
0.791s✔ Microsoft.Build.UnitTests.CopySingleThreaded_Tests   (41 tests)
1.382s✔ Microsoft.Build.UnitTests.CopySymbolicLink_Tests   (43 tests)

This is what I see in data for sample CI run after the change:

7.544s✔ Microsoft.Build.UnitTests.Copy_Tests   (133 tests)

So total number of tests went down - I suppose thanks to mentioned deduplication - great!
So I'd expect the total execution time to reflect that as well - instead it went up slightly (from 7.178 seconds to 7.544 seconds). 
While this is not concerning from the tests/CI run time perspective - I just want to be sure there is actually not any functional change causing a different scenario(s) to be run.

Once I'll get some time I'll try to dig to see if there is some specific test that have significantly different runtime on the both runs. If you want - you can do this by downloading one of the 'test logs' artifacts for this run and some other PR, extract the data from the xml and cross compare

image

Totally optional though!

Other than that the changes look good to go to me! I just want to have this extra precaution step before signing off.

File: src/Tasks.UnitTests/Copy_Tests.cs [hunk 1]
```diff
@@ -2101,130 +2151,66 @@ internal virtual void ErrorIfLinkFailedCheck()
...
-                    }
-                }
-
-                if (_throwOnFailure)
-                {
-                    throw new IOException("oops");
-                }
 
-                return null;
+                // TODO: Add test for MSB3892 when ErrorIfLinkFails && !UseHardlinksIfPossible && !UseSymboliclinksIfPossible
```
Comment (JanKrivanek) [inline src/Tasks.UnitTests/Copy_Tests.cs:2155]: You are welcome to create a work item for this.
It'll likely be marked for the contribution grabs though... but still more likely to be accomplished rather then code TODO

**Comment by jrdodds** [comment]
I started to look at some of the test logs. I compared logs from a build for this PR (8766) and from a build for PR 8763.

I compared "CoreOnWindows .Net 7.0 x64" and "CoreOnMac .Net 7.0 x64".

### CoreOnWindows .Net 7.0 x64

| PR | Tests | Time (s) | time/test |
| -- | ----- | ---- | ----- |
| 8763 | 211 | 7.965 | 0.0377 |
| 8766 | 122 | 6.048 | 0.0496 |

### CoreOnMac .Net 7.0 x64

| PR | Tests | Time (s) | time/test |
| -- | ----- | ---- | ----- |
| 8763 | 181 | 0.470 | 0.0026 |
| 8766 | 101 | 0.340 | 0.0496 |

This is far from exhaustive and controlled but the number of tests and the time both decreased as expected. The time per test, however, increased.

**Review by JanKrivanek — APPROVED** [review]
Thank you @jrdodds for the extra efforts.
Looking closer I confirm I actually see the total run time went down (I had an off-by-1 issue in my first quick look), comparing the actual testcases I can see that only the duplicated ones are droped.

Nice improvement!

**Review by AR-May — APPROVED** [review]
LGTM

**Pull request merged**